Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+
+* Eventual cardinality explosions have a big blast radius if a single Prometheus instance is responsible for scraping the majority of monitored applications.
+* Recoverying from crash takes several minutes due to WAL replay.
+
+Meanwhile, another strategy would be to use Horizontal Pod Autoscalers (HPAs) with Prometheus statefulsets and increase number of replicas. However, arguably it makes things more complicated:
+
+* Replicas with the same scrape configuration will not shard the scraping load into different instances. Instead, we get several Prometheus instances generating duplicated metrics.
+
+# Goals
+
+* Enable usage of HPAs to scale Prometheus up and down through sharding mechanisms.
+
+# Non-Goals
+
+* Implement our own autoscaling logic.
+
+# Audience
+
+* Users looking for a automated sharding mechanism with good cost-reliability ratio.
+
+# How
+
+Today, there are a few strategies to measure load on top of Prometheus instances.
+
+1. Perhaps you can use [Kubernetes Metrics API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api) to measure load in terms of CPU and/or Memory usage.
+
+2. Another strategy might be measure amount of samples ingested per second
+
+- Prometheus Agents: `rate(prometheus_agent_samples_appended_total[5m])`
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
+
+### Scale Status subresource
+
+When working with CRDs, common HPAs (e.g. Keda) depends on a Status subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Unfortunately, Prometheus-Operator resources still don't implement this field. There was an attempt to implement this in https://github.com/prometheus-operator/prometheus-operator/pull/4735, but was reverted because the PR was scaling replicas instead of shards. Although scaling replicas can help with High-Availability, it is expensive and hard to manage since it duplicates scrapes while not reducing the load on top of Prometheus because all replicas still scrape the same targets. Sharding serves as a better autoscaling strategy since the load is spread into all instances instead of getting duplicated.
+
+With only this change, Prometheus Agents can already be horizontally scaled without problems, but for Prometheus Servers it gets a little more complicated.
```

Comment:
I think that we still need to account for proper shutdown of the agents and ensure that the deletion happens only after the agent has forwarded all metrics to the remote write endpoints. It might be worth having a dedicated section on it. WDYT?

Reply:
I had the impression that Prometheus Agent would still try to finish the remote-write queue when receiving SIGTERM. I'll need to investigate

Reply:
Let's start with this assumption. We can adjust later if needed.